drivers/axi_core/spi_engine/spi_engine.c

```diff
@@ -0,0 +1,958 @@
+		}
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the spi engine
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param param Structure containing the spi init parameters
+ * @return int32_t - SUCCESS if the transfer finished
+ *		   - FAILURE if the memory allocation failed
+ */
+int32_t spi_engine_init(struct spi_desc **desc,
+			const struct spi_init_param *param)
+{
+	uint32_t			data_width;
+	uint32_t			spi_engine_version;
+	struct spi_desc			*local_desc;
+	struct spi_engine_desc		*eng_desc;
+	struct spi_engine_init_param	*spi_engine_init;
+
+	local_desc = (struct spi_desc *)malloc(sizeof(*local_desc));
+	eng_desc = (struct spi_engine_desc*)malloc(sizeof(*eng_desc));
+
+	if (!desc || !eng_desc)
+		return FAILURE;
+
+	spi_engine_init = param->extra;
+
+	local_desc->max_speed_hz = param->max_speed_hz;
+	local_desc->chip_select = param->chip_select;
+	local_desc->mode = param->mode;
+	local_desc->extra = eng_desc;
+
+	eng_desc->offload_config = OFFLOAD_DISABLED;
+	eng_desc->spi_engine_baseaddr = spi_engine_init->spi_engine_baseaddr;
+	eng_desc->type = spi_engine_init->type;
+	eng_desc->cs_delay = spi_engine_init->cs_delay;
+	eng_desc->clk_div =  SPI_ENGINE_AXI_SPEED_HZ /
+			     ((2 * param->max_speed_hz) - 1);
+
+	/* Perform a reset */
+	spi_engine_write(eng_desc, SPI_ENGINE_REG_RESET, 0x01);
+	usleep(1000);
+	spi_engine_write(eng_desc, SPI_ENGINE_REG_RESET, 0x00);
+
+	/* Get current data width */
+	spi_engine_read(eng_desc, SPI_ENGINE_REG_DATA_WIDTH, &data_width);
+	eng_desc->max_data_width = data_width;
+	dev_dbg(desc, "Data with is 8 bits (max %d)", eng_desc->max_data_width);
+	spi_engine_set_transfer_wdith(local_desc, spi_engine_init->data_width);
+
+	/* Get current data width */
+	spi_engine_read(eng_desc, SPI_ENGINE_REG_VERSION, &spi_engine_version);
+
+	*desc = local_desc;
+
+	dev_info(desc, "Spi engine v%d.%d.%d succesfully initialized at 0x%X.",
+		 (spi_engine_version >> 16),
+		 ((spi_engine_version >> 8) & 0xFF),
+		 (spi_engine_version & 0xFF),
+		 desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Write/read on the spi interface
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param data Pointer to data buffer
+ * @param bytes_number Number of bytes to transfer
+ * @return int32_t - SUCCESS if the transfer finished
+ *		   - FAILURE if the memory allocation or transfer failed
+ */
+int32_t spi_engine_write_and_read(struct spi_desc *desc,
+				  uint8_t *data,
+				  uint8_t bytes_number)
+{
+	uint8_t 		i;
+	uint8_t 		word_len;
+	uint8_t 		words_number;
+	int32_t 		ret;
+	struct spi_engine_msg	*msg;
+	struct spi_engine_desc	*desc_extra;
+
+	desc_extra = desc->extra;
+
+	words_number = spi_get_words_number(desc_extra, bytes_number);
+
+	msg = (spi_engine_msg *)malloc(sizeof(*msg));
+	msg->cmds = (spi_engine_cmd_queue*)malloc(sizeof(*msg->cmds));
+	if (!msg->cmds || !msg)
+		return FAILURE;
+
+	msg->tx_buf =(uint32_t*)malloc(words_number * sizeof(msg->tx_buf[0]));
```

Comment:
you can use calloc and it will fill the buffers with 0.
Also calloc it is prefered in all the allocations.